File: ft_transcendence/Frontend/js/user-data.js
function handle42Auth() {
	const timestamp = new Date().getTime();
	fetch('http://localhost:8000/api/42_intra_link', {
		method: 'GET',
		headers: {
			'Cache-Control': 'no-cache',
		},
	})
		.then(response => {
			if (!response.ok) {
				throw new Error('Network response was not ok');
			}
			return response.json();
		})
		.then(data => {
			console.log('Data fetched:', data);
			// Handle the fetched data
			forty_two_url = data.forty_two_url + `&state=${timestamp}` + `&random=${Math.random()}`;
			console.log(forty_two_url);
			window.location.href = forty_two_url;
		})
		.catch(error => {
			console.error('Error fetching data:', error);
		});

}

function handleLogout() {
	localStorage.clear();

	fetch('http://localhost:8000/logout', {
		credentials: 'include',
	})
		.then(response => {
			if (!response.ok) {
				throw new Error('Network response was not ok');
			}
			return response.json();
		})
		.then(data => {
			if (data.message === 'Logged out successfully') {
				document.getElementById('content').innerHTML = 'You have been logged out successfully';
			}
			else {
				document.getElementById('logout').remove();
			}
		})
		.catch(error => {
			console.error('Error fetching data:', error);
		});
}

document.addEventListener('DOMContentLoaded', () => {
	const dynamicLinks = document.querySelectorAll('.dynamic-link');

	dynamicLinks.forEach(link => {
		link.addEventListener('click', async event => {
			event.preventDefault();
			const route = event.target.getAttribute('href');
			console.log('Fetching data from:', route);
			await fetchBackendData(`http://localhost:8000/api/${route}/`)
				.then(data => {
					console.log('Data fetched:', data);
					updateDataContainer(data);
				})
		});
	});
});

async function fetchBackendData(route) {

	return await fetch(route, {
		headers: {
			'Authorization': `Bearer ${localStorage.getItem('token') ? localStorage.getItem('token') : localStorage.getItem('access_token')}`,
		},
	})
		.then(response => {
			if (!response.ok) {
				if (response.status === 401 || response.status === 403) {
					alert('Unauthorized');
				} else {
					// alert(`Network response was not ok ${response.status}`);
				}
			}
			return response.text();
		});
}

function updateDataContainer(data) {
	// Update the UI with the fetched data
	const dataContainer = document.getElementById('data-container');

	dataContainer.innerHTML = data;

}

function handleToken() {
	const credentials = {
		username : localStorage.getItem('username'),
		password : localStorage.getItem('username')
	};

	fetch('http://localhost:8000/api/token/', {
		method: 'POST',
		credentials: 'include',
		headers: {
			'Content-Type': 'application/json'
		},
		body: JSON.stringify(credentials),
	})
		.then(response => {
			if (!response.ok) {
				throw new Error('Network response was not ok');
			}
			return response.json();
		})
		.then(data => {
			localStorage.setItem('token', data.access);
			localStorage.setItem('refresh', data.refresh);
			console.log('Data fetched:', data);
		})
		.catch(error => {
			alert(error);
		});
}

File: user_data.c

async function handleAuth() {

	const urlParams = new URLSearchParams(window.location.search);
	const code = urlParams.get('code');
	var FORTY_TWO_CLIENT_ID = "u-s4t2ud-69155ca3ecf1f57fa6e8660a9988bbdd7f03a45128ea80a454d6f13939c4bca5"
	var FORTY_TWO_CLIENT_SECRET = "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"
	var FORTY_TWO_REDIRECT_URI = "https://localhost:8090/"
	console.log('code', code)
	if (code) {
		console.log("starting fetching....");
		await fetch(`http://localhost:8000/auth/?code=${code}`)
			.then(async response => {
				if (!response.ok) {
					alert('response was not ok');
				}
				data = await response.json();

				console.log('data', data)
				if (data.message) {
					console.log('message', message)
					return;
				}
				userToken = data.token;
				user = data.user;
				let csrfToken = data.csrfToken
				localStorage.setItem('access_token', userToken);
				localStorage.setItem('username', user.username);
				console.log(userToken);
				console.log(user);
				console.log(csrfToken);
				console.log(data.sessionId);
				// document.cookie = "sessionId=" + data.sessionId;
				console.log('authDone', authDone)
				const url = new URL(window.location.href);
				url.search = '';
				const mainUrl = url.toString();

				history.replaceState({}, '', mainUrl);






				const otp = data.otp
				console.log('otp', otp)
				if (otp === 'validate_otp') {
					console.log('validate otp');
					// setMainWindowframe();
					document.getElementById("content").innerHTML = `<div class="window-frame" id="main-window">
				<div class="top-bar">
				<img class="top-bar-child" alt="" src="./assets/public/rectangle-4.svg" />

				<div class="options">
				<img class="vector-icon" alt="" src="./assets/public/vector.svg" />

				<img class="dot-grid-icon" alt="" src="./assets/public/dot-grid.svg" />
				</div>
				</div>
				<div class="window"></div>
				</div>`;
					document.getElementsByClassName('window')[0].innerHTML = `
				<div class="mb-3 p-20px">
				<label for="otp-input" class="form-label">OTP Code</label>
				<input type="text" class="form-control" id="otp-input" placeholder="Enter OTP code">
				</div>
				<button type="submit-otp" id="submit-otp" class="btn btn-primary">Validate OTP</button>
				`;

					document.getElementById('submit-otp').addEventListener('click', async () => {
						let otp = document.getElementById('otp-input').value;
						if (!otp) {
							alert('Please enter OTP code');
							return;
						}
						const requestBody = new URLSearchParams();
						requestBody.append('username', urlUsername);
						requestBody.append('otp', otp);

						await fetch('http://localhost:8000/validate_otp/', {
							method: 'POST',
							headers: {
								'Content-Type': 'application/x-www-form-urlencoded',
								'Authorization': `Bearer ${token}`,
								'x-csrftoken': getCookie('csrftoken'),
							},
							body: requestBody.toString(),
						})
							.then(response => {
								if (!response.ok) {
									throw new Error(response.statusText === 'Unauthorized' ? 'Unauthorized' : 'Network response was not ok');
								}
								return response.json();
							})
							.then(data => {
								// Handle the response data here
								if (data.message === 'OTP is valid') {
									console.log(data);
									localStorage.setItem('access_token', token);
									document.getElementsByClassName("window")[0].innerHTML = '';
									alert('OTP is valid, enjoy pongos');
									return false;
								}
								else {
									alert('Invalid OTP code');
								}

							})
							.catch(error => {
								console.error('Error:', error);
							});
					});
				}

			})
	}

}

handleAuth();